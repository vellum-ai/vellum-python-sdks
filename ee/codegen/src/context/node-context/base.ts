import { VellumVariableType } from "vellum-ai/api";

import {
  VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH,
  VELLUM_WORKFLOW_NODES_MODULE_PATH,
} from "src/constants";
import { WorkflowContext } from "src/context";
import { PortContext } from "src/context/port-context";
import { NodeOutputNotFoundError } from "src/generators/errors";
import { NodePort as NodePortType, WorkflowDataNode } from "src/types/vellum";
import { toValidPythonIdentifier } from "src/utils/casing";
import { getNodeLabel } from "src/utils/nodes";
import {
  doesModulePathStartWith,
  getGeneratedNodeDisplayModulePath,
  getGeneratedNodeModuleInfo,
} from "src/utils/paths";
import { hasErrorOutput } from "src/utils/typing";

export declare namespace BaseNodeContext {
  interface Args<T extends WorkflowDataNode> {
    workflowContext: WorkflowContext;
    nodeData: T;
    importOrder: readonly number[];
  }
}

export abstract class BaseNodeContext<T extends WorkflowDataNode> {
  public readonly workflowContext: WorkflowContext;
  public readonly importOrder: readonly number[];
  public readonly nodeModulePath: string[];
  public readonly nodeModuleName: string;
  public readonly nodeFileName: string;
  public readonly nodeClassName: string;
  public readonly nodeDisplayModulePath: string[];
  public readonly nodeDisplayModuleName: string;
  public readonly nodeDisplayFileName: string;
  public readonly nodeDisplayClassName: string;

  public nodeData: T;
  public abstract readonly baseNodeClassName: string;
  public abstract readonly baseNodeDisplayClassName: string;

  protected isCore = false;
  public readonly baseNodeClassModulePath: readonly string[];
  public readonly baseNodeDisplayClassModulePath: readonly string[];

  private nodeOutputNamesById: Record<string, string> | undefined;
  public readonly portContextsById: Map<string, PortContext>;
  public readonly defaultPortContext: PortContext | undefined;

  private readonly nodePortNames: Set<string> = new Set();

  constructor(args: BaseNodeContext.Args<T>) {
    this.workflowContext = args.workflowContext;
    this.nodeData = args.nodeData;
    this.importOrder = args.importOrder;

    if (args.nodeData.type === "GENERIC") {
      if (
        !doesModulePathStartWith(
          args.nodeData.base?.module,
          VELLUM_WORKFLOW_BASE_NODES_MODULE_PATH
        )
      ) {
        this.baseNodeClassModulePath = args.nodeData.base.module;
      } else {
        this.baseNodeClassModulePath = VELLUM_WORKFLOW_NODES_MODULE_PATH;
      }
    } else if (
      !doesModulePathStartWith(
        args.nodeData.base?.module,
        VELLUM_WORKFLOW_NODES_MODULE_PATH
      )
    ) {
      this.baseNodeClassModulePath =
        args.nodeData.base?.module ??
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    } else {
      this.baseNodeClassModulePath =
        this.workflowContext.sdkModulePathNames.DISPLAYABLE_NODES_MODULE_PATH;
    }
    this.baseNodeDisplayClassModulePath =
      this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH;

    const { moduleName, nodeClassName, modulePath } =
      getGeneratedNodeModuleInfo({
        workflowContext: args.workflowContext,
        nodeDefinition: args.nodeData.definition,
        nodeLabel: this.getNodeLabel(),
      });
    this.nodeModuleName = moduleName;
    this.nodeClassName = nodeClassName;
    this.nodeModulePath = modulePath;

    this.workflowContext.addUsedClassName(nodeClassName);
    this.nodeFileName = `${this.nodeModuleName}.py`;

    this.nodeDisplayModuleName = this.nodeModuleName;
    this.nodeDisplayFileName = `${this.nodeDisplayModuleName}.py`;
    this.nodeDisplayClassName = `${this.nodeClassName}Display`;
    this.nodeDisplayModulePath = getGeneratedNodeDisplayModulePath(
      args.workflowContext,
      this.nodeDisplayModuleName
    );

    const portContexts = this.nodeData.ports
      ? this.createPortContextsFromNodePortAttribute(this.nodeData.ports)
      : this.createPortContexts();
    portContexts.forEach((portContext) => {
      this.workflowContext.addPortContext(portContext);
    });

    this.portContextsById = new Map(
      portContexts.map((portContext) => [portContext.portId, portContext])
    );

    this.defaultPortContext = portContexts.find(
      (portContext) => portContext.isDefault
    );
  }

  protected abstract getNodeOutputNamesById(): Record<string, string>;
  protected abstract getNodeOutputTypesById(): Record<
    string,
    VellumVariableType
  >;
  protected abstract createPortContexts(): PortContext[];

  protected createPortContextsFromNodePortAttribute(
    nodePorts: NodePortType[]
  ): PortContext[] {
    return nodePorts.map(
      (port) =>
        new PortContext({
          workflowContext: this.workflowContext,
          nodeContext: this,
          portId: port.id,
          portName: port.name,
          isDefault: port.type === "DEFAULT",
        })
    );
  }

  public getNodeLabel(): string {
    return getNodeLabel(this.nodeData);
  }

  public getNodeOutputNameById(outputId: string): string | undefined {
    // Lazily load node output names
    if (!this.nodeOutputNamesById) {
      this.nodeOutputNamesById = this.getNodeOutputNamesById();
    }

    const nodeOutputName = this.nodeOutputNamesById[outputId];
    if (!nodeOutputName) {
      // This handles an edge case where the node references a non-existent subworkflow deployment
      if (
        this.nodeData.type === "SUBWORKFLOW" &&
        this.nodeData.data.variant === "DEPLOYMENT"
      ) {
        this.workflowContext.addError(
          new NodeOutputNotFoundError(
            `Could not find Subworkflow Deployment Output with id ${outputId}`,
            "WARNING"
          )
        );
        return;
      }

      this.workflowContext.addError(
        new NodeOutputNotFoundError(
          `Failed to find ${this.nodeClassName} Output with id '${outputId}'`,
          "WARNING"
        )
      );
      return;
    }

    return toValidPythonIdentifier(nodeOutputName, "output");
  }

  public getNodeOutputTypeById(
    outputId: string
  ): VellumVariableType | undefined {
    const nodeOutputTypesById = this.getNodeOutputTypesById();

    return nodeOutputTypesById[outputId];
  }

  public isPortNameUsed(portName: string): boolean {
    return this.nodePortNames.has(portName);
  }

  public addUsedPortName(portName: string): void {
    this.nodePortNames.add(portName);
  }

  /**
   * Extend this class to perform any additional property generation asynchronously
   * after the node context has been created.
   */
  public buildProperties(): Promise<void> {
    return Promise.resolve();
  }

  public isImportedBefore(
    nodeContext: BaseNodeContext<WorkflowDataNode>
  ): boolean {
    for (
      let i = 0;
      i < Math.max(this.importOrder.length, nodeContext.importOrder.length);
      i++
    ) {
      const thisImportOrder = this.importOrder[i];
      const otherImportOrder = nodeContext.importOrder[i];
      if (thisImportOrder === undefined && otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder === undefined) {
        return true;
      }

      if (otherImportOrder === undefined) {
        return false;
      }

      if (thisImportOrder < otherImportOrder) {
        return true;
      }

      if (thisImportOrder > otherImportOrder) {
        return false;
      }
    }
    return false;
  }

  protected getErrorOutputId(): string | undefined {
    const TRY_NODE_NAME = "TryNode";
    const tryNodes = this.nodeData.adornments?.filter((adornment) => {
      return adornment.base.name === TRY_NODE_NAME;
    });

    const tryNode = tryNodes?.at(-1);

    let fallbackErrorOutputId: string | undefined = undefined;
    if (hasErrorOutput(this.nodeData)) {
      fallbackErrorOutputId = this.nodeData.data.errorOutputId;
    }

    return tryNode ? tryNode.id : fallbackErrorOutputId;
  }
}
