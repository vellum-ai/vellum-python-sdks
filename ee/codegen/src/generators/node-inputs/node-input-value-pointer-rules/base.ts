import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  IterableConfig,
  NodeInputValuePointerRule as NodeInputValuePointerRuleType,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace BaseNodeInputValuePointerRule {
  export interface Args<T extends NodeInputValuePointerRuleType> {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: T;
    iterableConfig?: IterableConfig;
  }
}

export abstract class BaseNodeInputValuePointerRule<
  T extends NodeInputValuePointerRuleType
> extends AstNode {
  private readonly nodeContext: BaseNodeContext<WorkflowDataNode>;
  public readonly workflowContext: WorkflowContext;
  public readonly nodeInputValuePointerRule: T;
  public readonly iterableConfig?: IterableConfig;
  private astNode: AstNode | undefined;

  constructor({
    nodeContext,
    nodeInputValuePointerRule,
    iterableConfig,
  }: BaseNodeInputValuePointerRule.Args<T>) {
    super();
    this.nodeContext = nodeContext;
    this.workflowContext = nodeContext.workflowContext;
    this.iterableConfig = iterableConfig;
    this.nodeInputValuePointerRule = nodeInputValuePointerRule;

    this.astNode = this.getAstNode();
    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  public getReferencedNodeContext():
    | BaseNodeContext<WorkflowDataNode>
    | undefined {
    return undefined;
  }

  abstract getAstNode(): AstNode | undefined;

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
