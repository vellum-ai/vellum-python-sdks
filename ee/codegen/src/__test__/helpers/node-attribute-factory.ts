import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
