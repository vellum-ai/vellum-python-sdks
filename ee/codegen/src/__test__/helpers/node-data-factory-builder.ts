import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
