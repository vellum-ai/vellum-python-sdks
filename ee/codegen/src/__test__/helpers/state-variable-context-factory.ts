import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
