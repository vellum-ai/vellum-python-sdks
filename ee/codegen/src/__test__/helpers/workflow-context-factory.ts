import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
