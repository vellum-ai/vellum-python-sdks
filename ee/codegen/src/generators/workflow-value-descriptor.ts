import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { isNil } from "lodash";

import { WorkflowContext } from "src/context";
import {
  NodeAttributeGenerationError,
  NodePortGenerationError,
} from "src/generators/errors";
import { Expression } from "src/generators/expression";
import { WorkflowValueDescriptorReference } from "src/generators/workflow-value-descriptor-reference/workflow-value-descriptor-reference";
import {
  IterableConfig,
  OperatorMapping,
  WorkflowValueDescriptor as WorkflowValueDescriptorType,
} from "src/types/vellum";
import { assertUnreachable } from "src/utils/typing";

export namespace WorkflowValueDescriptor {
  export interface Args {
    workflowValueDescriptor: WorkflowValueDescriptorType;
    workflowContext: WorkflowContext;
    iterableConfig?: IterableConfig;
  }
}

export class WorkflowValueDescriptor extends AstNode {
  private workflowContext: WorkflowContext;
  private iterableConfig?: IterableConfig;
  private astNode: AstNode;

  public constructor(args: WorkflowValueDescriptor.Args) {
    super();

    this.workflowContext = args.workflowContext;
    this.iterableConfig = args.iterableConfig;
    this.astNode = this.generateWorkflowValueDescriptor(
      args.workflowValueDescriptor
    );
    this.inheritReferences(this.astNode);
  }

  private generateWorkflowValueDescriptor(
    workflowValueDescriptor: WorkflowValueDescriptorType
  ): AstNode {
    if (isNil(workflowValueDescriptor)) {
      return python.TypeInstantiation.none();
    }
    return this.buildExpression(workflowValueDescriptor);
  }

  private buildExpression(
    workflowValueDescriptor: WorkflowValueDescriptorType
  ): AstNode {
    // Base case
    if (this.isReference(workflowValueDescriptor)) {
      return new WorkflowValueDescriptorReference({
        workflowContext: this.workflowContext,
        workflowValueReferencePointer: workflowValueDescriptor,
        iterableConfig: this.iterableConfig,
      });
    }

    switch (workflowValueDescriptor.type) {
      case "UNARY_EXPRESSION": {
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        const operator = this.convertOperatorType(workflowValueDescriptor);
        return new Expression({
          lhs,
          operator: operator,
        });
      }
      case "BINARY_EXPRESSION": {
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        const rhs = this.buildExpression(workflowValueDescriptor.rhs);
        const operator = this.convertOperatorType(workflowValueDescriptor);
        return new Expression({
          lhs,
          operator: operator,
          rhs,
        });
      }
      case "TERNARY_EXPRESSION": {
        const base = this.buildExpression(workflowValueDescriptor.base);
        const lhs = this.buildExpression(workflowValueDescriptor.lhs);
        const rhs = this.buildExpression(workflowValueDescriptor.rhs);
        const operator = this.convertOperatorType(workflowValueDescriptor);
        return new Expression({
          lhs: lhs,
          operator: operator,
          rhs: rhs,
          base: base,
        });
      }
      default:
        assertUnreachable(workflowValueDescriptor);
    }
  }

  private convertOperatorType(
    workflowValueDescriptor: WorkflowValueDescriptorType
  ): OperatorMapping {
    if (this.isExpression(workflowValueDescriptor)) {
      const operator = workflowValueDescriptor.operator;
      const operatorMappings: Record<string, OperatorMapping> = {
        "=": "equals",
        "!=": "does_not_equal",
        "<": "less_than",
        ">": "greater_than",
        "<=": "less_than_or_equal_to",
        ">=": "greater_than_or_equal_to",
        contains: "contains",
        beginsWith: "begins_with",
        endsWith: "ends_with",
        doesNotContain: "does_not_contain",
        doesNotBeginWith: "does_not_begin_with",
        doesNotEndWith: "does_not_end_with",
        null: "is_null",
        notNull: "is_not_null",
        in: "in",
        notIn: "not_in",
        between: "between",
        notBetween: "not_between",
      };
      const value = operatorMappings[operator];
      if (!value) {
        throw new NodePortGenerationError(
          `This operator: ${operator} is not supported`
        );
      }
      return value;
    } else {
      throw new NodeAttributeGenerationError(
        `Operators should exist on expression and not be null`
      );
    }
  }

  private isExpression(workflowValueDescriptor: WorkflowValueDescriptorType) {
    return (
      workflowValueDescriptor.type === "UNARY_EXPRESSION" ||
      workflowValueDescriptor.type === "BINARY_EXPRESSION" ||
      workflowValueDescriptor.type === "TERNARY_EXPRESSION"
    );
  }

  private isReference(workflowValueDescriptor: WorkflowValueDescriptorType) {
    return (
      workflowValueDescriptor.type === "NODE_OUTPUT" ||
      workflowValueDescriptor.type === "WORKFLOW_INPUT" ||
      workflowValueDescriptor.type === "WORKFLOW_STATE" ||
      workflowValueDescriptor.type === "CONSTANT_VALUE" ||
      workflowValueDescriptor.type === "VELLUM_SECRET" ||
      workflowValueDescriptor.type === "EXECUTION_COUNTER"
    );
  }

  write(writer: Writer): void {
    this.astNode.write(writer);
  }
}
