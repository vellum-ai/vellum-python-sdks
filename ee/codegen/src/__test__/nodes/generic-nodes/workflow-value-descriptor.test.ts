import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach, describe, expect, it } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import { WorkflowValueDescriptor } from "src/generators/workflow-value-descriptor";
import {
  WorkflowDataNode,
  WorkflowValueDescriptor as WorkflowValueDescriptorType,
} from "src/types/vellum";

describe("WorkflowValueDescriptor", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "input-1",
          key: "count",
          type: "NUMBER",
        },
        workflowContext,
      })
    );

    vi.spyOn(workflowContext, "getNodeContext").mockReturnValue({
      nodeClassName: "TestNode",
      path: ["nodes", "test-node-path"],
      getNodeOutputNameById: vi.fn().mockReturnValue("my_output"),
    } as unknown as BaseNodeContext<WorkflowDataNode>);
  });

  describe("expressions", () => {
    it("generates unary expression with input variable", async () => {
      const descriptor: WorkflowValueDescriptorType = {
        type: "UNARY_EXPRESSION",
        operator: "null",
        lhs: {
          type: "WORKFLOW_INPUT",
          inputVariableId: "input-1",
        },
      };

      const valueDescriptor = new WorkflowValueDescriptor({
        workflowValueDescriptor: descriptor,
        workflowContext,
      });

      valueDescriptor.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("generates binary expression with node output and constant", async () => {
      const descriptor: WorkflowValueDescriptorType = {
        type: "BINARY_EXPRESSION",
        operator: "=",
        lhs: {
          type: "NODE_OUTPUT",
          nodeId: "node-1",
          nodeOutputId: "output-1",
        },
        rhs: {
          type: "CONSTANT_VALUE",
          value: {
            type: "STRING",
            value: "expected-value",
          },
        },
      };

      const valueDescriptor = new WorkflowValueDescriptor({
        workflowValueDescriptor: descriptor,
        workflowContext,
      });

      valueDescriptor.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("generates binary expression with nested workflow value descriptors", async () => {
      const descriptor: WorkflowValueDescriptorType = {
        type: "BINARY_EXPRESSION",
        operator: "=",
        lhs: {
          type: "BINARY_EXPRESSION",
          operator: "=",
          lhs: {
            type: "NODE_OUTPUT",
            nodeId: "node-1",
            nodeOutputId: "output-1",
          },
          rhs: {
            type: "CONSTANT_VALUE",
            value: {
              type: "STRING",
              value: "expected-value",
            },
          },
        },
        rhs: {
          type: "CONSTANT_VALUE",
          value: {
            type: "STRING",
            value: "another-expected-value",
          },
        },
      };

      const valueDescriptor = new WorkflowValueDescriptor({
        workflowValueDescriptor: descriptor,
        workflowContext,
      });

      valueDescriptor.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("generates ternary expression with input variables", async () => {
      const descriptor: WorkflowValueDescriptorType = {
        type: "TERNARY_EXPRESSION",
        operator: "between",
        base: {
          type: "WORKFLOW_INPUT",
          inputVariableId: "input-1",
        },
        lhs: {
          type: "CONSTANT_VALUE",
          value: {
            type: "NUMBER",
            value: 1,
          },
        },
        rhs: {
          type: "CONSTANT_VALUE",
          value: {
            type: "NUMBER",
            value: 10,
          },
        },
      };

      const valueDescriptor = new WorkflowValueDescriptor({
        workflowValueDescriptor: descriptor,
        workflowContext,
      });

      valueDescriptor.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
