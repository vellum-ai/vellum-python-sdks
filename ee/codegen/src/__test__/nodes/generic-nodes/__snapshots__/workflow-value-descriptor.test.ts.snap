// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkflowValueDescriptor > expressions > generates binary expression with nested workflow value descriptors 1`] = `
"TestNode.Outputs.my_output.equals("expected-value").equals("another-expected-value")
"
`;

exports[`WorkflowValueDescriptor > expressions > generates binary expression with node output and constant 1`] = `
"TestNode.Outputs.my_output.equals("expected-value")
"
`;

exports[`WorkflowValueDescriptor > expressions > generates ternary expression with input variables 1`] = `
"Inputs.count.between(1, 10)
"
`;

exports[`WorkflowValueDescriptor > expressions > generates unary expression with constant value for parse_json 1`] = `
"ConstantValueReference('{"key": "value"}').parse_json()
"
`;

exports[`WorkflowValueDescriptor > expressions > generates unary expression with input variable 1`] = `
"Inputs.count.is_null()
"
`;

exports[`WorkflowValueDescriptor > expressions > generates unary expression with is_error operator 1`] = `
"TestNode.Outputs.my_output.is_error()
"
`;

exports[`WorkflowValueDescriptor > expressions that begin with constant values > generates binary expression beginning with constant value reference 1`] = `
"ConstantValueReference("Hello, World!").equals(Inputs.count)
"
`;

exports[`WorkflowValueDescriptor > expressions that begin with constant values > generates ternary expression beginning with constant value reference 1`] = `
"ConstantValueReference(123).between(Inputs.count, Inputs.another_count)
"
`;

exports[`WorkflowValueDescriptor > expressions that begin with constant values > generates unary expression beginning with constant value reference 1`] = `
"ConstantValueReference("Hello, World!").is_null()
"
`;

exports[`WorkflowValueDescriptor > expressions that begin with constant values > generates unary expression with is_error operator on constant value 1`] = `
"ConstantValueReference("Some error message").is_error()
"
`;
